onnxscript/rewriter/ort_fusions/_core.py

```diff
@@ -26,9 +32,32 @@
     *fused_matmul_rule_sets.fused_matmul_rule_sets(),
 ]
 
+_extra_opt_rules = llama_p0_rule_set()
 
-def fuse_xformers(model: ir.Model) -> None:
+
+# Preliminary optimizations before applying the transformer fusions.
+# TODO: There are some potential redundancies below. Can be targeted for optimization
+# once we have robust fusion.
+def _pre_optimize(model):
+    optimize(model)
+    # TODO: Do we need this dependence on ONNX's partial-data-propagation? There are some
+    # extra shape-propagation and partial-data-propagation rules in ONNX that are not yet
+    # incorporated in our optimizer.
+    with tempfile.TemporaryDirectory() as tmpdir:
```

Comment:
Is it possible to instead run shape inference on a model without initializers, then re-assemble it into an ir Model? This way we don’t need to write to disk

Reply:
We can create a shape inference pass that does this

Reply:
https://github.com/microsoft/onnxscript/pull/2117